README.md
# VNC Reflector

## About VNC Reflector

VNC Reflector is a specialized VNC server which acts as a proxy
sitting between real VNC server (host) and a number of VNC clients. It
was designed to work efficiently with large number of clients, but
current version is optimized to work primarily with clients that
request 8-bit color format known as BGR233. Other color formats are
supported too but performance may be worse if there are many such
clients.

VNC Reflector can switch between different host connections on the
fly, supports reverse host connections and hosts with variable desktop
geometries. It supports full-control and read-only client connections
depending on which of two passwords was used to initiate a client
session.

Raw, Hextile, Tight and CopyRect encodings are suported both at the
host side and at the client side. JPEG sub-encoding in Tight is
supported on client connections.

Please note that the documentation is incomplete.


## Licensing terms

* Copyright (C) 2017 alejandro_liu@hotmail.com
* Copyright (C) 2001-2003 HorizonLive.com, Inc.  
* Copyright (c) 1988,1989,1990,1991,1992 by Richard Outerbridge.  
  (GEnie : OUTER; CIS : [71755,204]) Graven Imagery, 1992.
* Copyright 1987, 1988, 1989, 1998  The Open Group

All rights reserved.

This software is released under the terms specified in the file LICENSE,
included.

This software was originally authored by Constantin Kaplinsky <const@ce.cctpu.edu.ru>
and sponsored by HorizonLive.com, Inc.

## Command-line usage

* * *
````
./vncreflector [OPTIONS...] HOST_INFO_FILE
````
* * *



Options:

*  -i PID_FILE     - write pid file, appending listening port to the filename
*  -p PASSWD_FILE  - read a plaintext client password file [default: passwd]
*  -a ACTIVE_FILE  - create file during times when a host is connected
*  -c ACTIONS_FILE - on events, execute commands specified in a file
*  -l LISTEN_PORT  - port to listen for client connections [default: 5999]
*  -b IP_ADDRESS   - bind listening sockets to a specific IP [default: any]
*  -s FBS_PREFIX   - save host sessions in rfbproxy-compatible files
                    (optionally appending 3-digit session IDs to the
                    filename prefix, only if used without the -j option)
*  -j              - join saved sessions (see -s option) in one session file
*  -t              - use Tight encoding for host communications if possible
*  -T COMPR_LEVEL  - like -t, but use the specified compression level (1..9)
*  -g LOG_FILE     - write logs to the specified file [default: reflector.log]
*  -v LOG_LEVEL    - set verbosity level for the log file (0..6) [default: 4]
*  -f LOG_LEVEL    - run in foreground, show logs on stderr at the specified
                    verbosity level (0..6) [note: use 3 for normal output]
*  -q              - suppress printing copyright banner at startup
*  -h              - print this help message

## VNC Multiplexer mode

If the `LISTEN_PORT` specified is `0` (zero), then VNC multiplexer mode
is activated.  What this does is that instead of listening on a port
it will simply use `STDIN/STDOUT` as a pre-opened socket.  As if it was
invoked from [inetd][inetd].  Then the `HOST_INFO_FILE` is treated
as a script that will launch a `vnc` session (again in [inetd][inetd]
mode.  The script will have an argument which is the file descriptor
that will receive a new host to connect to in the format of the
`HOST_INFO_FILE` explained later.

## Format of the HOST_INFO_FILE

HOST_INFO_FILE provides information about the host to connect to. It's
first line should begin with hostname (or with an asterisk if we would wait
for reversed host->reflector connection). Next thing is VNC display number
after a colon (or a TCP port number, for reversed connections). Then
follows plaintext host password after a space. For non-reversed
connections, negative display numbers are allowed (e.g. display number
-5820 corresponds to port number 80 at the host side). For reversed
connections, default listening port number is 5500.

A sample to connect to some.hostname.com's display :0 (port 5900):

* * *
````
some.hostname.com:0 PaS$w0rD
````
* * *

A sample to listen for host connections on TCP port 5598:

* * *
````
*:5598 PaS$w0rD
````
* * *

## Format of the PASSWD_FILE

PASSWD_FILE keeps passwords for client connections.

It may contain two lines with one password on each line. First line is for
full-control client password, second line is for read-only client password.
Empty lines denote no authentication. If the file contains one line, then
there is no read-only password, and all clients would have been granted
full-control access.


## Format of the ACTIONS_FILE

ACTIONS_FILE used with the -c option specifies commands that should be
executed upon certain events. Each line starts with a keyword denoting a
particular event type, then follows colon character (":"), and after an
optional sequence of spaces follows the command to execute. An example:

* * *
````
host_activate: echo "new host" >> /tmp/connections.txt
````
* * *

Currently, the only supported event type is `host_activate', which is raised
on each new host connection.


## Control signals

  * SIGHUP signal causes VNC Reflector to disconnect all connected clients.

  * SIGUSR1 signal tells VNC Reflector that it should re-read
    HOST_INFO_FILE and to reconnect to host specified there (or to start
    listening for reversed host connections). Host connection would be
    closed when new host connection is established successfully (auth is ok
    etc.). Client connections would be preserved.

  * SIGUSR2 signal tells VNC Reflector that it should close current host
    connection, then re-read HOST_INFO_FILE and to reconnect to host
    specified there (or to start listening for reversed host connections). 
    Client connections would be preserved.

## TODO

1. Enable Client SetDesktopSize message
   - host_io: Implement ExtendedDesktopSize pseudo encoding support
   - client_io: Implement SetDesktopSize message handling (sending
     error)
   - client_io: Implement ExtendedDesktopSize semantics
   - client_io: Implement SetDesktopSize message handling
2. display reader code:
   - upon receiving reconnect message
   - try connect to new host
   - if success, we switch to new host and set the host
     slot to quit program on close.
   - if fail, we continue with the old host.  Old host should
     have the loginscript running and after a while if it stays
     running it assumes that something went wrong.  Show a message
     and allow to re-login.
3. In inetd mode, when entering inetd mode, we set the client slot
   to quit program on close.  The host slot should be setup on close
   to quit program.  (We could re-run loginscript but most likely
   that would lead to an infinite loop)
4. Add TLS support to client connection
   - Note, the *Free* VNC applet in Chrome doesn't support TLS!
   - Would not work with Protocol v3.3

    [inetd]: https://en.wikipedia.org/wiki/Inetd "About Inetd"


ChangeLog
2017-11:

	Added inetd mode/vnc multiplexer mode.
	Pseudo-encoding for New Desktop name.

2012-11-06 04:10 r10r

	Applied patches from fedora upstream
	http://pkgs.fedoraproject.org/cgit/vnc-reflector.git
	* Fixes a logging crash when running on foreground
	* Fixes a one-byte buffer overflow in formatting RFB version string


2003-04-27 02:47  const

	* async_io.c: The aio_mainloop() version which uses select() did
	not install signal handlers.

2003-04-22 00:20  const

	* README, actions.c, active.c, async_io.c, async_io.h, client_io.h,
	control.c, decode_hextile.c, decode_tight.c, encode.c, encode.h,
	encode_tight.c, fbs_files.c, host_connect.h, host_io.c, host_io.h,
	logging.c, logging.h, reflector.h, rfblib.c, translate.c,
	translate.h: Version strings changed for version 1.2.4, copyrights
	updated.

2003-04-17 00:32  const

	* Makefile: Build with speed optimizations by default.

2003-04-17 00:31  const

	* encode_tight.c: Major speed optimization in the CheckSolidTile()
	function, made after profiling the reflector with many client
	connections active.

2003-04-10 01:44  const

	* Makefile, client_io.c, region.h, region_more.c: Trying to reduce
	the number of rectangles in regions before sending updates to
	clients.

2003-01-11 15:44  const

	* not-connected.pnm: File removed.

2003-01-11 15:44  const

	* README, client_io.c, host_connect.c, main.c, reflector.h,
	rfblib.h: Version strings changed for the 1.2.3 release, copyright
	strings updated.

2003-01-11 15:14  const

	* rfblib.h: Removed a non-sense "FIXME" comment.

2003-01-11 15:13  const

	* host_connect.c: Minor cleanup.

2003-01-09 21:35  const

	* client_io.c: Finished full support for CopyRect encoding. Fixed a
	number of minor issues in client update handling.

2003-01-09 14:47  const

	* rfblib.h: One more "FIXME" comment.

2003-01-09 14:47  const

	* client_io.c: Fixed a logical bug in tracking client update
	requests -- the reflector could stop sending updates if changes in
	the framebuffer did not affect the rectangle requested by the
	client.

2003-01-09 12:41  const

	* client_io.c: Clearing pending CopyRect regions on receiving
	non-incremental update requests from clients.

2003-01-09 12:25  const

	* client_io.c: Fixes and cleanups in handling of pending client
	updates.

2003-01-06 13:07  const

	* client_io.c: NewFBSize handling has been improved.

2003-01-06 10:38  const

	* Makefile: Added -L/usr/local/lib to LDFLAGS.

2002-10-11 19:05  const

	* client_io.c: Making sure rectangle dimensions in updates will fit
	into area requested by clients. Not a fully reliable solution yet,
	but should work in real-life cases.

2002-10-03 23:25  const

	* client_io.c, client_io.h, host_io.c: First implementation of
	support for CopyRect encoding in new X regions based architecture.

2002-09-30 12:11  const

	* client_io.c, client_io.h: Fixed NewFBSize handling which was
	seriously broken after porting X regions code.

2002-09-22 17:40  const

	* decode_tight.c: Fixed a problem with resetting zlib streams on
	host switching in saved sessions.

2002-09-21 20:08  const

	* decode_tight.c, host_io.c, host_io.h: Bugfix -- zlib streams were
	not reset between host connections.

2002-09-21 19:43  const

	* Makefile, client_io.c, client_io.h, control.c, decode_hextile.c,
	decode_tight.c, encode.c, encode_tight.c, host_connect.c,
	host_io.c, main.c, rect.c, rect.h, region.c, region.h, rfblib.h,
	translate.c: Ported X region stuff, from miregion.c. This will
	allow the Reflector to maintain not just a list of rectangles in
	the same order as received from the host, but an X region structure
	which consists of non-overlapped rectangles.

2002-09-18 18:01  const

	* README, reflector.h: Version strings changed, version 1.2.2.

2002-09-18 18:01  const

	* decode_hextile.c, decode_tight.c, fbs_files.c, reflector.h: Now
	Tight-encoded host sessions can be saved in FBS files.	Minor
	optimization in writing FBS sessions.

2002-09-10 14:59  const

	* README, reflector.h: Version strings changed, version 1.2.1.

2002-09-10 14:42  const

	* encode_tight.c: Decreased JPEG quality factors for low quality
	levels.

2002-09-10 14:10  const

	* client_io.c: Fixed a bug causing double free() when a client
	requests change in its pixel format.

2002-09-10 13:58  const

	* encode_tight.c: Fixed a bug with color translation into the
	24-bit mode in solid-color areas.

2002-09-09 14:35  const

	* LICENSE, README, async_io.c, async_io.h, client_io.c,
	client_io.h, control.c, decode_hextile.c, decode_tight.c, encode.c,
	encode.h, fbs_files.c, host_connect.c, host_connect.h, host_io.c,
	host_io.h, logging.c, logging.h, rect.c, rect.h, reflector.h,
	rfblib.c, rfblib.h, translate.c, translate.h: Copyright strings
	updated.

2002-09-09 14:27  const

	* README, reflector.h: Version strings changed, version 1.2.0.

2002-09-09 02:37  const

	* Makefile, README, client_io.c, client_io.h, encode.c, encode.h,
	encode_tight.c, encode_tight8.c, rect.h, rfblib.h: Implemented
	fully-functional Tight encoder with JPEG compression.

2002-09-07 13:10  const

	* README, host_connect.c, host_connect.h, main.c: New -t and -T
	command-line options to enable Tight encoding and to specify
	compression level. Now the default encoding is Hextile again.

2002-09-04 10:07  const

	* encode.c, encode.h: It was not necessary to make the Hextile
	cache variables global -- now they're static.

2002-09-04 09:53  const

	* decode_tight.c: Implemented the "gradient" filter in the Tight
	decoder.  Better error checking.

2002-09-04 09:52  const

	* rfblib.h: Tiny cleanup.

2002-09-04 09:12  const

	* decode_tight.c: Fixed two problems introduced in the previous
	commit.

2002-09-04 09:09  const

	* decode_tight.c: Implemented checking for errors where necessary.
	Removed debug output.  Minor optimizations.

2002-09-04 02:57  const

	* Makefile, decode_hextile.c, decode_tight.c, encode_tight8.c,
	host_connect.c, host_io.c, host_io.h, main.c, rfblib.h,

main.c
/* VNC Reflector
 * Copyright (C) 2017 alejandro_liu@hotmail.com.  All rights reserved.g
 * Copyright (C) 2001-2003 HorizonLive.com, Inc.  All rights reserved.
 *
 * This software is released under the terms specified in the file LICENSE,
 * included.  HorizonLive provides e-Learning and collaborative synchronous
 * presentation solutions in a totally Web-based environment.  For more
 * information about HorizonLive, please see our website at
 * http://www.horizonlive.com.
 *
 * This software was authored by Constantin Kaplinsky <const@ce.cctpu.edu.ru>
 * and sponsored by HorizonLive.com, Inc.
 *
 * $Id: main.c,v 1.47 2003/01/11 09:44:02 const Exp $
 * Main module
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <zlib.h>

#include "rfblib.h"
#include "async_io.h"
#include "logging.h"
#include "reflector.h"
#include "host_connect.h"
#include "translate.h"
#include "host_io.h"
#include "client_io.h"
#include "encode.h"
#include "utils.h"

/*
 * Configuration options
 */

static int   opt_no_banner;
static int   opt_cl_listen_port;
static char *opt_log_filename;
static char *opt_passwd_filename;
static char *opt_active_filename;
static char *opt_actions_filename;
static int   opt_foreground;
static int   opt_stderr_loglevel;
static int   opt_file_loglevel;
static char  opt_pid_file[256];
static char *opt_fbs_prefix;
static int   opt_join_sessions;
static char *opt_bind_ip;
static int   opt_request_tight;
static int   opt_tight_level;

static unsigned char opt_client_password[9];
static unsigned char opt_client_ro_password[9];

static char *opt_host_info_file;

/*
 * Global variables
 */

RFB_SCREEN_INFO g_screen_info;
CARD32 *g_framebuffer;
CARD16 g_fb_width, g_fb_height;

/*
 * Functions local to this file
 */

static void parse_args(int argc, char **argv);
static void report_usage(char *program_name);
static int read_password_file(void);
static int init_screen_info(void);
static int write_pid_file(void);
static int remove_pid_file(void);

/*
 * Implementation
 */

int main(int argc, char **argv)
{
  long cache_hits, cache_misses;

  init_stdio();
  /* Parse command line, exit on error */
  parse_args(argc, argv);

  if (!opt_no_banner) {
    fprintf(stderr,
"VNC Reflector %s.  Copyright (C) 2001-2003 HorizonLive.com, Inc.\n\n"
"HorizonLive provides e-Learning and collaborative synchronous presentation\n"
"solutions in a totally Web-based environment.  For more information about\n"
"HorizonLive, please see our website at http://www.horizonlive.com/\n\n",
            VERSION);
  }

  if (!log_open(opt_log_filename, opt_file_loglevel,
                (opt_foreground) ? opt_stderr_loglevel : -1)) {
    fprintf(stderr, "%s: error opening log file (ignoring this error)\n",
            argv[0]);
  }

  log_write(LL_MSG, "Starting VNC Reflector %s", VERSION);

  /* Fork the process to the background if necessary */
  if (!opt_foreground) {
    if (!opt_no_banner) {
      fprintf(stderr, "Starting in the background, "
              "see the log file for errors and other messages.\n");
    }

    if (getpid() != 1) {
      signal(SIGTTIN, SIG_IGN);
      signal(SIGTTOU, SIG_IGN);
      signal(SIGTSTP, SIG_IGN);
      if (fork ())
        return 0;
      setsid();
    }
    close(0);
    close(1);
    close(2);
    log_write(LL_INFO, "Switched to the background mode");
  }

  /* Initialization */
  if (init_screen_info()) {
    read_password_file();
    set_host_encodings(opt_request_tight, opt_tight_level);
    set_client_passwords(opt_client_password, opt_client_ro_password);
    fbs_set_prefix(opt_fbs_prefix, opt_join_sessions);

    set_active_file(opt_active_filename);
    set_actions_file(opt_actions_filename);

    aio_init();
    if (opt_bind_ip != NULL) {
      if (aio_set_bind_address(opt_bind_ip)) {
        log_write(LL_INFO, "Would bind listening sockets to address %s",
                  opt_bind_ip);
      } else {
        log_write(LL_WARN, "Illegal address to bind listening sockets to: %s",
                  opt_bind_ip);
      }
    }

    /* Main work */
    if (opt_cl_listen_port == 0) {
      set_client_listen_port(opt_cl_listen_port);
      if (connect_script(opt_host_info_file)) {
	aio_mainloop();
      }
    } else {
      if (connect_to_host(opt_host_info_file, opt_cl_listen_port)) {
	if (write_pid_file()) {
	  set_control_signals();
	  aio_mainloop();
	  remove_pid_file();
	}
      }
    }

    /* Cleanup */
    if (g_framebuffer != NULL) {
      log_write(LL_DETAIL, "Freeing framebuffer and associated structures");
      free(g_framebuffer);
      free_enc_cache();
    }
    if (g_screen_info.name != NULL)
      free(g_screen_info.name);

    get_hextile_caching_stats(&cache_hits, &cache_misses);
    if (cache_hits + cache_misses != 0) {
      log_write(LL_INFO, "Hextile BGR233 caching efficiency: %d%%",
                (int)((cache_hits * 100 + (cache_hits + cache_misses) / 2)
                      / (cache_hits + cache_misses)));
    }
  }

  log_write(LL_MSG, "Terminating");

  /* Close logs */
  if (!log_close() && opt_foreground) {
    fprintf(stderr, "%s: error closing log file (ignoring this error)\n",
            argv[0]);
  }

  /* Done */
  exit(1);
}

static void parse_args(int argc, char **argv)
{
  int err = 0;
  int c;
  char *temp_pid_file = NULL;
  char temp_buf[32];            /* 32 bytes should be more than enough */

  opt_no_banner = 0;
  opt_foreground = 0;
  opt_stderr_loglevel = -1;
  opt_file_loglevel = -1;
  opt_passwd_filename = NULL;
  opt_log_filename = NULL;
  opt_active_filename = NULL;
  opt_cl_listen_port = -1;
  opt_pid_file[0] = '\0';
  opt_fbs_prefix = NULL;
  opt_join_sessions = 0;
  opt_bind_ip = NULL;
  opt_request_tight = 0;
  opt_tight_level = -1;

  while (!err &&
         (c = getopt(argc, argv, "hqjv:f:p:a:c:g:l:i:s:b:tT:")) != -1) {
    switch (c) {
    case 'h':
      err = 1;
      break;
    case 'q':
      opt_no_banner = 1;
      break;
    case 'j':
      opt_join_sessions = 1;
      break;
    case 'v':
      if (opt_file_loglevel != -1)
        err = 1;
      else
        opt_file_loglevel = atoi(optarg);
      break;
    case 'f':
      opt_foreground = 1;
      if (opt_stderr_loglevel != -1)
        err = 1;
      else
        opt_stderr_loglevel = atoi(optarg);
      break;
    case 'p':
      if (opt_passwd_filename != NULL)
        err = 1;
      else
        opt_passwd_filename = optarg;
      break;
    case 'g':
      if (opt_log_filename != NULL)
        err = 1;
      else
        opt_log_filename = optarg;
      break;
    case 'a':
      if (opt_active_filename != NULL)
        err = 1;
      else
        opt_active_filename = optarg;
      break;
    case 'c':
      if (opt_actions_filename != NULL)
        err = 1;
      else
        opt_actions_filename = optarg;
      break;
    case 'l':
      if (opt_cl_listen_port != -1)
        err = 1;
      else {
        opt_cl_listen_port = atoi(optarg);
        if (opt_cl_listen_port < 0)
          err = 1;
      }
      break;
    case 'i':
      if (temp_pid_file != NULL)
        err = 1;
      else
        temp_pid_file = optarg;
      break;
    case 's':
      if (opt_fbs_prefix != NULL)
        err = 1;
      else
        opt_fbs_prefix = optarg;
      break;
    case 'b':
      if (opt_bind_ip != NULL)
        err = 1;
      else
        opt_bind_ip = optarg;
      break;
    case 't':
      if (opt_request_tight)
        err = 1;
      else
        opt_request_tight = 1;
      break;
    case 'T':
      if (opt_request_tight) {
        err = 1;
      } else {
        opt_request_tight = 1;
        opt_tight_level = atoi(optarg);
        if (opt_tight_level <= 0 || opt_tight_level > 9)
          err = 1;
      }
      break;
    default:
      err = 1;
    }
  }

  /* Print usage help on error */
  if (err || optind != argc - 1) {
    report_usage(argv[0]);
    exit(1);
  }

  /* Provide reasonable defaults for some options */
  if (opt_file_loglevel == -1)
    opt_file_loglevel = LL_INFO;
  if (opt_passwd_filename == NULL)
    opt_passwd_filename = "passwd";
  if (opt_log_filename == NULL)
    opt_log_filename = "reflector.log";
  if (opt_cl_listen_port == -1)
    opt_cl_listen_port = 5999;

  /* Append listening port number to pid filename */
  if (temp_pid_file != NULL) {
    sprintf(temp_buf, "%d", opt_cl_listen_port);
    sprintf(opt_pid_file, "%.*s.%s", (int)(255 - strlen(temp_buf) - 1),
            temp_pid_file, temp_buf);
  }

  /* Save pointer to host info filename */
  opt_host_info_file = argv[optind];
}

static void report_usage(char *program_name)
{
  fprintf(stderr,
          "VNC Reflector %s.  Copyright (C) 2001-2003 HorizonLive.com, Inc."
          "\n\n",
          VERSION);

  fprintf(stderr, "Usage: %s [OPTIONS...] HOST_INFO_FILE\n\n",
          program_name);

  fprintf(stderr,
          "Options:\n"
          "  -i PID_FILE     - write pid file, appending listening port"
          " to the filename\n"
          "  -p PASSWD_FILE  - read a plaintext client password file"
          " [default: passwd]\n"
          "  -a ACTIVE_FILE  - create file during times when a host is"
          " connected\n"
          "  -c ACTIONS_FILE - on events, execute commands specified in"
          " a file\n"
          "  -l LISTEN_PORT  - port to listen for client connections"
          " [default: 5999]\n"
          "  -b IP_ADDRESS   - bind listening sockets to a specific IP"
          " [default: any]\n");
  fprintf(stderr,
          "  -s FBS_PREFIX   - save host sessions in rfbproxy-compatible"
          " files\n"
          "                    (optionally appending 3-digit session IDs"
          " to the\n"
          "                    filename prefix, only if used without the"
          " -j option)\n"
          "  -j              - join saved sessions (see -s option) in one"
          " session file\n"
          "  -t              - use Tight encoding for host communications"
          " if possible\n"
          "  -T COMPR_LEVEL  - like -t, but use the specified compression"
          " level (1..9)\n");
  fprintf(stderr,
          "  -g LOG_FILE     - write logs to the specified file"
          " [default: reflector.log]\n"
          "  -v LOG_LEVEL    - set verbosity level for the log file (0..%d)"
          " [default: %d]\n"
          "  -f LOG_LEVEL    - run in foreground, show logs on stderr"
          " at the specified\n"
          "                    verbosity level (0..%d) [note: use %d for"
          " normal output]\n"
          "  -q              - suppress printing copyright banner at startup\n"
          "  -h              - print this help message\n\n",
          LL_DEBUG, LL_INFO, LL_DEBUG, LL_MSG);

  fprintf(stderr,
          "Please refer to the README file for a description of the file"
          " formats for\n"
          "  HOST_INFO_FILE and PASSWD_FILE files mentioned above in this"
          " help text.\n\n");
}

static int read_password_file(void)
{
  FILE *passwd_fp;
  unsigned char *password_ptr = opt_client_password;
  int line = 0, len = 0;
  int c;

  /* Fill passwords with zeros */
  memset(opt_client_password, 0, 9);
  memset(opt_client_ro_password, 0, 9);

  log_write(LL_DETAIL, "Looking for passwords in the file \"%s\"",
            opt_passwd_filename);

  passwd_fp = fopen(opt_passwd_filename, "r");
  if (passwd_fp == NULL) {
    log_write(LL_WARN,
              "No client password file, assuming no authentication");
    return 1;
  }

  /* Read password file */
  while (line < 2) {
    c = getc(passwd_fp);
    if (c == '\r')
      c = getc(passwd_fp);      /* Handle MS-DOS-style end of line */
    if (c != '\n' && c != EOF && len < 8) {
      password_ptr[len++] = c;
    } else {
      password_ptr[len] = '\0';
      /* Truncate long passwords */
      if (len == 8 && c != '\n' && c != EOF) {
        log_write(LL_WARN, "Using only 8 first bytes of a longer password");
        do {
          c = getc(passwd_fp);
        } while (c != '\n' && c != EOF);
      }
      /* End of file */
      if (c == EOF)
        break;
      /* Empty password means no authentication */
      if (len == 0) {
        log_write(LL_WARN, "Got empty client password, hoping no auth is ok");
      }
      /* End of line */
      if (++line == 1) {
        password_ptr = opt_client_ro_password;
      }
      len = 0;
    }
  }
  if (len == 0) {
    if (line == 0) {
      log_write(LL_WARN, "Client password not specified, assuming no auth");
    } else {
      line--;
    }
  }

  log_write(LL_DEBUG, "Got %d password(s) from file, including empty ones",
            line + 1);

  /* Provide reasonable defaults if not all two passwords set */
  if (line == 0) {
    log_write(LL_DETAIL, "Read-only client password not specified");
    strcpy((char *)opt_client_ro_password, (char *)opt_client_password);
  }

  fclose(passwd_fp);
  return 1;
}

static int init_screen_info(void)
{
  union _LITTLE_ENDIAN {
    CARD32 value32;
    CARD8 test;
  } little_endian;

  /* Set the initial desktop name */
  g_screen_info.name_length = 1;
  g_screen_info.name = malloc(2);
  if (g_screen_info.name == NULL) {
    log_write(LL_ERROR, "Error allocating memory");
    return 0;
  }
  memcpy(g_screen_info.name, "?", 2);

  /* Fill in the PIXEL_FORMAT structure */
  g_screen_info.pixformat.bits_pixel = 32;
  g_screen_info.pixformat.color_depth = 24;
  g_screen_info.pixformat.true_color = 1;
  g_screen_info.pixformat.r_max = 255;
  g_screen_info.pixformat.g_max = 255;
  g_screen_info.pixformat.b_max = 255;
  g_screen_info.pixformat.r_shift = 16;
  g_screen_info.pixformat.g_shift = 8;
  g_screen_info.pixformat.b_shift = 0;

  /* Set correct endian flag in PIXEL_FORMAT */
  little_endian.value32 = 1;
  if (little_endian.test) {
    log_write(LL_DEBUG, "Our machine is little endian");
    g_screen_info.pixformat.big_endian = 0;
  } else {
    log_write(LL_DEBUG, "Our machine is big endian");
    g_screen_info.pixformat.big_endian = 1;
  }

  /* Make sure we would not try to free the framebuffer */
  g_framebuffer = NULL;

  return 1;
}

static int write_pid_file(void)
{
  int pid_fd, len;
  char buf[32];                 /* 32 bytes should be more than enough */

  if (opt_pid_file[0] == '\0')
    return 1;

  pid_fd = open(opt_pid_file, O_WRONLY | O_CREAT | O_EXCL, 0644);
  if (pid_fd == -1) {
    log_write(LL_ERROR, "Pid file exists, another instance may be running");
    return 0;
  }
  sprintf(buf, "%d\n", (int)getpid());
  len = strlen(buf);
  if (write(pid_fd, buf, len) != len) {
    close(pid_fd);
    log_write(LL_ERROR, "Error writing to pid file");
    return 0;
  }

  log_write(LL_DEBUG, "Wrote pid file: %s", opt_pid_file);

  close(pid_fd);
  return 1;
}

static int remove_pid_file(void)
{
  if (opt_pid_file[0] != '\0') {
    if (unlink(opt_pid_file) == 0) {
      log_write(LL_DEBUG, "Removed pid file", opt_pid_file);
    } else {
      log_write(LL_WARN, "Error removing pid file: %s", opt_pid_file);
      return 0;
    }
  }
  return 1;
}

